include/swift/Basic/ExperimentalDependencies.h

```diff
@@ -0,0 +1,968 @@
+  /// fingerprint *after* the node has been created.
+  void setFingerprint(Optional<std::string> fp) {
+    fingerprint = fp;
+    assert(ensureThatTheFingerprintIsValidForSerialization());
+  }
+
+  const Optional<std::string> &getSwiftDeps() const {
+    ensureThatTheSwiftDepsIsValidForSerialization();
+    return swiftDeps;
+  }
+
+  /// Nodes can move from file to file when the driver reads the result of a
+  /// compilation.
+  void setSwiftDeps(Optional<std::string> s) {
+    swiftDeps = s;
+    ensureThatTheSwiftDepsIsValidForSerialization();
+  }
+
+  void dump() const;
+
+  bool assertImplementationsMustBeInFiles() const {
+    assert((getSwiftDeps().hasValue() || !getKey().isImplementation()) &&
+           "Implementations must be in some file.");
+    return true;
+  }
+
+  std::string humanReadableName() const;
+
+protected:
+  /// Convert to- or from- another format via the argument functions.
+  /// See \ref FrontendNode::serializeOrDeserialize.
+  void serializeOrDeserialize(
+      function_ref<void(size_t &)> convertInt,
+      function_ref<void(std::string &)> convertString,
+      function_ref<void(Optional<std::string> &)> convertOptionalString);
+
+private:
+  bool ensureThatTheFingerprintIsValidForSerialization() const;
+  bool ensureThatTheSwiftDepsIsValidForSerialization() const;
+};
+
+//==============================================================================
+// MARK: FrontendNode
+//==============================================================================
+
+class FrontendGraph;
+
+/// A node in a graph that represents the dependencies computed when compiling a
+/// single primary source file. Such a graph is always constructed
+/// monotonically; it never shrinks or changes once finished. The frontend does
+/// not need to be able to remove nodes from the graph, so it can represent arcs
+/// with a simple format relying on sequence numbers.
+class FrontendNode : public Node {
+  /// To represent Arcs in a serializable fashion, the code puts all nodes in
+  /// the graph in a vector (`allNodes`), and uses the index in that vector to
+  /// refer to the node.
+  size_t sequenceNumber = ~0;
+
+  /// Holds the sequence numbers of my uses.
+  std::unordered_set<size_t> usesOfMe;
+
+public:
+  /// When the driver imports a node create an uninitialized instance for
+  /// deserializing.
+  FrontendNode() : Node(), sequenceNumber(~0) {}
+
+  /// Used by the frontend to build nodes.
+  FrontendNode(DependencyKey key, Optional<std::string> fingerprint)
+      : Node(key, fingerprint) {
+    assert(key.verify());
+  }
+
+  bool operator==(const FrontendNode &other) const {
+    return Node::operator==(other) && sequenceNumber == other.sequenceNumber &&
+           usesOfMe == other.usesOfMe;
+  }
+
+
+  /// True iff this node represents a decl in this file or the file itself.
+  /// False iff this node represents a use from another file.
+  bool isInSameFile() const { return getSwiftDeps().hasValue(); }
```

Comment:
I forgot this was a *Frontend*Node and so this method name didn't make sense at first. Could have used more of a reminder.

Reply:
Expanding my thought: now that we have batch mode, one frontend job can handle multiple files. Maybe this structure should be called SingleFileDependencyGraph or something?

Reply:
The comment is easily clarified. You have a point about the graph name, too. But then what's a good node name? `SingleFileGraphNode`? Seems confusing. Maybe `FrontEndSingleFileDependencyGraph`? Long! I *could* just point out the single-file-ness in a comment on `FrontendGraph`... Will sleep on it.